sqllogictest-bin/src/main.rs

```diff
@@ -645,19 +601,127 @@ async fn flush(out: &mut impl std::io::Write) -> std::io::Result<()> {
+                case
+            }
+            RunResult::Err(e) => {
+                let mut status = TestCaseStatus::non_success(NonSuccessKind::Failure);
+                status.set_type("test failure");
+
+                let mut case = TestCase::new(test_case_name, status);
+                case.set_system_err(e.to_string());
+                case.set_time(Duration::from_millis(0));
+                case.set_system_out("");
+                case.set_timestamp(Local::now());
+                case.set_classname(junit);
+                case
+            }
+            RunResult::Skipped | RunResult::Cancelled => {
+                // TODO: what status should we use for cancelled tests?
+                let mut case = TestCase::new(test_case_name, TestCaseStatus::skipped());
+                case.set_time(Duration::from_millis(0));
+                case.set_timestamp(Local::now());
+                case.set_classname(junit);
+                case
+            }
+        }
+    }
+}
+
 async fn connect_and_run_test_file(
     out: &mut impl std::io::Write,
     filename: PathBuf,
     engine: &EngineConfig,
     config: DBConfig,
     labels: &[String],
-) -> Result<Duration> {
+    cancel: CancellationToken,
+) -> RunResult {
+    static RUNNING_TESTS: tokio::sync::RwLock<()> = tokio::sync::RwLock::const_new(());
+
+    // If the run is already cancelled, skip it.
+    if cancel.is_cancelled() {
+        // Ensure that all running tests are cancelled before we return `Skipped`.
+        let _ = RUNNING_TESTS.write().await;
+
+        writeln!(
+            out,
+            "{: <60} .. {}",
+            filename.to_string_lossy(),
+            style("[SKIPPED]").dim().bold(),
+        )
+        .unwrap();
+        return RunResult::Skipped;
+    }
+
+    // Hold until the current test is finished or cancelled.
+    let _running = RUNNING_TESTS.read().await;
+
     let mut runner = Runner::new(|| engines::connect(engine, &config));
     for label in labels {
         runner.add_label(label);
     }
     runner.set_var(well_known::DATABASE.to_owned(), config.db.clone());
-    let result = run_test_file(out, &mut runner, filename).await;
+
+    let begin = Instant::now();
+
+    // Note: we don't use `CancellationToken::run_until_cancelled` here because it always
+    // poll the wrapped future first, while we want cancellation to be more responsive.
+    let result = tokio::select! {
+        biased;
+        _ = cancel.cancelled() => {
+            writeln!(
+                out,
+                "{} after {} ms",
+                style("[CANCELLED]").yellow().bold(),
+                begin.elapsed().as_millis(),
+            )
+            .unwrap();
+            RunResult::Cancelled
+        }
+        result = run_test_file(out, &mut runner, filename) => {
+            if let Err(err) = &result {
+                writeln!(
+                    out,
+                    "{} after {} ms\n\n{:?}\n",
```

Comment:
Previously this was done by duplicated logic in `run_serial` and `run_parallel`. Also there's no running time shown.